main.c

```diff
@@ -943,6 +943,27 @@ static int tcmu_set_max_fd_limit(const int nr_files)
 	return 0;
 }
 
+
+static struct tcmu_config* tcmu_initiatize_options(void)
+{
+	struct tcmu_config *cfg;
+	char *log_dir;
+
+	cfg = calloc(1, sizeof(*cfg));
+	if (cfg == NULL) {
+		tcmu_err("allocating TCMU config failed: %m\n");
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	log_dir = getenv("TCMU_LOGDIR");
+	snprintf(cfg->def_log_dir, PATH_MAX, log_dir ? log_dir : TCMU_LOG_DIR_DEFAULT);
+	cfg->def_log_level = TCMU_CONF_LOG_INFO;
+	snprintf(cfg->path, PATH_MAX, TCMU_CONFIGFILE_DEFAULT);
```

Comment:
I think we want to rename this to tcmu_initialize_config() and put it in libtcmu_config.c. libtcmu_config.c will then manage both the allocation and freeing of the tcmu_config object.

Also in main() with your changes we can fix the ordering so the cleanup/freeing path is the reverse of the setup/allocation path right?

main()

tcmu_initialize_config()
tcmu_load_config()
tcmu_setup_log()
tcmu_watch_config()

....

tcmu_unwatch_config()
tcmu_destroy_log()
tcmu_free_config()

so then that matches the error unwinding path too.

Reply:
Makesense done, I will update the series with a new patch about organizing the cleanup order